app/controllers/concerns/two_factor_authenticatable.rb

```diff
@@ -78,9 +81,19 @@ def render_show_after_invalid
 
   def update_invalid_user
     current_user.second_factor_attempts_count += 1
-    # set time lock if max attempts reached
-    current_user.second_factor_locked_at = Time.zone.now if current_user.max_login_attempts?
-    current_user.save
+
+    UpdateUser.new(
+      user: current_user,
+      attributes: set_time_lock_if_max_attempts_reached
```

Comment:
I think it is. I read it like this:
1. Attributes are being set. Some are set conditionally.
2. Update the user with the final attributes.

This way, I can see all the attributes in one place, and the update is performed using one method.

In your version, we are mixing and matching 2 ways to update the user. First, we use the setter, then we use `update`. I don't think I've seen that combination before. I've seen either setter + `#save`, or `update` by itself. I think it has a higher cognitive load. I read it like this:
1. Set some attributes on the user via setter
2. Call `UpdateUser` with some other attributes
3. Scroll down to see what those attributes are. Notice that they can be empty, which first made me think, why are we making a DB call with empty attributes? Then I realized there were attributes being set previously via the setter call.